### PR TITLE
fix($sanitize): use appropriate inert document strategy for Firefox and Safari

### PR DESCRIPTION
**Pull request by petebacondarwin**
Both Firefox and Safari are vulnerable to XSS if we use an inert document
created via `document.implementation.createHTMLDocument()`.

Now we check for those vulnerabilities and then use a DOMParser or XHR
strategy if needed.

Thanks to @cure53 for the heads up on this issue.

### CODE REVIEW

**Comment by cure53** [comment]
:+1:

File: src/ngSanitize/sanitize.js [hunk 1]
```diff
@@ -313,16 +313,72 @@ function $SanitizeProvider() {
...
+      xhr.responseType = 'document';
+      xhr.open('GET', 'data:text/html;charset=utf-8,' + html, false);
+      xhr.send(null);
+      var body = xhr.response.body;
+      body.firstChild.remove();
+      return body;
+    }
+
+    function getInertBodyElement_DOMParser(html) {
+      html = '<remove></remove>' + html;
```
Comment (mgol) [inline src/ngSanitize/sanitize.js]: Why is that needed?

Comment (petebacondarwin) [inline src/ngSanitize/sanitize.js]: One moment :-)

Comment (petebacondarwin) [inline src/ngSanitize/sanitize.js]: I added a comment!

**Review by gkalpak — APPROVED** [review]

File: test/ngSanitize/sanitizeSpec.js [hunk 2]
```diff
@@ -266,14 +268,25 @@ describe('HTML', function() {
     });
   });
 
+  // See https://github.com/cure53/DOMPurify/blob/a992d3a75031cb8bb032e5ea8399ba972bdf9a65/src/purify.js#L439-L449
```
Comment (mgol) [inline test/ngSanitize/sanitizeSpec.js]: Could you add support comments? Here:
```js
// Support: Safari 10.1+
```
and below:
```js
// Support: Firefox 53+
```

Comment (petebacondarwin) [inline test/ngSanitize/sanitizeSpec.js]: Since they are feature sniffing, I was not sure if we needed to mention the type and version of the browser... I expect that Firefox do not consider their part a bug, which means that it will never be resolved. But in any case, there may be other browsers out there that exhibit the same behaviour.

Is the aim of these comments to indicate when we can remove it?

Comment (mgol) [inline test/ngSanitize/sanitizeSpec.js]: Yes, it's mostly to know if/when to remove it. But I commented in a wrong place, the tests should stay forever as they are, such comments belong more in source.

Comment (petebacondarwin) [inline test/ngSanitize/sanitizeSpec.js]: I added comments here: https://github.com/angular/angular.js/pull/16010/files#diff-699beb0d8c48a8257d3d70331bf6138dR319

**Comment by koto** [comment]
LGTM.

**Comment by rjamet** [comment]
LGTM too.

**Comment by petebacondarwin** [comment]
Cherry-picked to 1.6.x
